util/Hash.h

```diff
@@ -15,16 +15,100 @@
 #ifndef Hash_H
 #define Hash_H
 
+/*
+ * SipHash is a family of pseudorandom functions (a.k.a. keyed hash functions)
+ * optimized for speed on short messages.
+ *
+ * Target applications include network traffic authentication and defense against
+ * hash-flooding DoS attacks.
+ *
+ * Solution inspired by code from:
+ *     Jean-Philippe Aumasson (https://131002.net/siphash/siphash24.c)
+ * */
+
 #include <stdint.h>
+typedef uint64_t u64;
+typedef uint32_t u32;
+typedef uint8_t   u8;
+
+#define Hash_ROTL(x,b) (u64)( ((x) << (b)) | ( (x) >> (64 - (b))) )
 
-/** The DJB2a hash (equivilant to DJB2 but using XOR instead of +) */
-static uint32_t Hash_compute(uint8_t* str, int length)
+#define Hash_U8TO64_LE(p) \
+    (((u64)((p)[0])      ) | \
+     ((u64)((p)[1]) <<  8) | \
+     ((u64)((p)[2]) << 16) | \
+     ((u64)((p)[3]) << 24) | \
+     ((u64)((p)[4]) << 32) | \
+     ((u64)((p)[5]) << 40) | \
+     ((u64)((p)[6]) << 48) | \
+     ((u64)((p)[7]) << 56))
+
+#define Hash_SIPROUND            \
+    do {              \
+        v0 += v1; v1=Hash_ROTL(v1,13); v1 ^= v0; v0=Hash_ROTL(v0,32); \
+        v2 += v3; v3=Hash_ROTL(v3,16); v3 ^= v2;                      \
+        v0 += v3; v3=Hash_ROTL(v3,21); v3 ^= v0;                      \
+        v2 += v1; v1=Hash_ROTL(v1,17); v1 ^= v2; v2=Hash_ROTL(v2,32); \
+    } while (0);
+
+/* SipHash-2-4 */
+static uint64_t Hash_compute(const uint8_t* in, int length, uint8_t key[16])
 {
-    uint32_t hash = 5381;
-    for (int i = 0; i < length; i++) {
-        hash = ((hash << 5) + hash) ^ str[i];
+    /* "somepseudorandomlygeneratedbytes" */
+    u64 v0 = 0x736f6d6570736575ULL;
+    u64 v1 = 0x646f72616e646f6dULL;
+    u64 v2 = 0x6c7967656e657261ULL;
+    u64 v3 = 0x7465646279746573ULL;
+    u64 b;
```

Comment:
better uint64_t because it's used everywhere else, if u64 is not available on some weird system this is the only file which will fail to compile.